storage: call EnsureSafeSplitKey during load-based splits

tbg opened the pull request:
We could end up splitting between column families of the same row,
which is illegal. Unfortunately the KV layer has to uphold invariants
here that it doesn't quite have introspection into, but after this
commit it hopefully stops breaking them.

See https://github.com/cockroachdb/cockroach/issues/16344 for some
additional history.

Possibly the solution for #39794.
Possibly the solution for #36834.
Possibly the solution for #36356.

(Intentionally not closing the above; leaving that to the SQL folks).

Closes #42056 (which is the go-to for reading up on this issue).

Release note (bug fix): prevent a number of panics from the SQL layer
caused by an invalid range split. These would usually manifest with
messages mentioning encoding errors ("found null on not null column" but
also possibly various others).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42833)
<!-- Reviewable:end -->

tbg committed:
storage: call EnsureSafeSplitKey during load-based splits

We could end up splitting between column families of the same row,
which is illegal. Unfortunately the KV layer has to uphold invariants
here that it doesn't quite have introspection into, but after this
commit it hopefully stops breaking them.

See https://github.com/cockroachdb/cockroach/issues/16344 for some
additional history.

Possibly the solution for #39794.
Possibly the solution for #36834.
Possibly the solution for #36356.

(Intentionally not closing the above; leaving that to the SQL folks).

Closes #42056 (which is the go-to for reading up on this issue).

Release note (bug fix): prevent a number of panics from the SQL layer
caused by an invalid range split. These would usually manifest with
messages mentioning encoding errors ("found null on not null column" but
also possibly various others).

tbg commented:
bors r=bdarnell

CI failed on a logic test for which I've pinged the issue

tbg commented:
[12:28:36]
[Test web UI] TOTAL: 1 FAILED, 317 SUCCESS
[12:28:36]
[Test web UI] Makefile:1323: recipe for target 'ui-test' failed

#42365 

bors r=bdarnell

tbg commented:
bors r=bdarnell

sigh

knz commented:
lemme take this over the finish line for you

bors r=bdarnell,a=tbg

knz commented:
Bors is unhappy because the CLA assistant didn't validate this. A no-op push would fix it.

tbg commented:
Thanks @knz.

bors r=bdarnell

tbg committed:
Makefile: skip ui-test

This can be re-enabled after
https://github.com/cockroachdb/cockroach/issues/42365 is fixed.

Release note: None

tbg commented:
Both CIs failed on `make ui-test` so I went ahead and skipped that; it's been a repeat offender.

bors r=bdarnell
🙏
